enclave-runner/src/usercalls/interface.rs

```diff
@@ -12,150 +12,278 @@ use std::slice;
 use fortanix_sgx_abi::*;
 
 use super::abi::{UsercallResult, Usercalls};
-use super::{EnclaveAbort, RunningTcs};
+use super::{EnclaveAbort, IOHandlerInput};
+use crate::futures::FutureExt;
+use futures::future::Future;
 
-pub(super) struct Handler<'a>(pub &'a mut RunningTcs);
+pub(super) struct Handler<'ioinput, 'tcs>(pub &'ioinput mut IOHandlerInput<'tcs>);
 
-impl<'a> Usercalls for Handler<'a> {
+impl<'future, 'ioinput: 'future, 'tcs: 'ioinput> Usercalls<'future> for Handler<'ioinput, 'tcs> {
     fn is_exiting(&self) -> bool {
         self.0.is_exiting()
     }
 
-    fn read(&mut self, fd: Fd, buf: *mut u8, len: usize) -> UsercallResult<(Result, usize)> {
-        unsafe {
-            Ok(from_raw_parts_mut_nonnull(buf, len)
-                .and_then(|buf| self.0.read(fd, buf))
-                .to_sgx_result())
+    fn read(
```

Comment:
It's unfortunate how much more verbose these implementations have become, mainly because `?` can no longer be used. I tried writing some macros to make it nicer, but they all failed. Maybe @jseyfried has some ideas.